Add analyzer to warn users when they reference a Workspace API from a…

mavasani opened the pull request:
…n Analyzer

Fixes #1588

AArnott commented:
This looks good. A couple of questions though based on the violations I hit that I want to be sure this analyzer covers:

1. Use of extension methods defined in the workspaces assembly
1. Use of `Simplifier` type to annotate `TypeSyntax`

It could be that your code already ensures these will be found. I couldn't find extension methods tested though, so I wanted to make sure. Thanks.

AArnott commented:
Also how about type checks and casts--is that covered?

BTW, I think the ability to find transitive uses of types is fascinating. I may be able to use that to enhance our threading analyzers to better detect when the UI thread is required transitively from a synchronous method.

mavasani commented:
Thanks, I was only computing explicit references, it won’t work for extension methods - good catch. I will enhance the analyzer to also look at implicit references. I will also add tests for type checks and casts, that should work with current implementation.

heejaechang commented on src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAPIUsageAnalyzer.cs:
```diff
@@ -0,0 +1,200 @@
+
+                Func<SyntaxNode, bool> descendantIntoChildren = node => ReferenceEquals(node, syntax) || !IsNamedTypeDeclarationBlock(node);
+                var semanticModel = compilation.GetSemanticModel(syntax.SyntaxTree);
+                foreach (TTypeSyntax typeSyntax in syntax.DescendantNodes(descendantIntoChildren).OfType<TTypeSyntax>())
```
can we use IOperation instead of using Syntax?

heejaechang replied:
shouldn't this check all expression not just type syntax?

heejaechang replied:
so, since it is checking all executable code, can we get Ioperation tree for each code block, (of diagnostic analzyer) and visit all Ioperation and see its type is from those workspace assemblies?

mavasani replied:
@heejaechang Thank you! I figured out that we need to visit operation tree for implicit type references, which was not covered by my previous implementation. However, for executable code too there are bunch of type syntax nodes for which we have no operations. For example, VB method block returns the operation tree, but there is no operation for type syntax nodes for parameters or return type. Lambda and local function declarations do not have specific operation nodes for parameters or return type either.

So for simplicity, I have taken the following approach:
1. Visit all operations in executable code and handle `IOperation.Type` for these nodes - no need to handle individual `OperationKind` (we need to handle static member accesses specially as they have no instance operation).
2. Visit all explicit `TypeSyntax` descendants in the named type declaration - covers types explicitly mentioned in declarations and executable code.

heejaechang commented on src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAPIUsageAnalyzer.cs:
```diff
@@ -0,0 +1,200 @@
+                    var violatingUsedTypeNamesBuilder = new SortedSet<string>();
+                    foreach (INamedTypeSymbol declaredType in namedTypesToAccessedTypesMap.Keys)
+                    {
+                        if (!diagnosticAnalyzerTypes.Contains(declaredType))
```
is this possible?

mavasani replied:
Yes, for helper types defined in assembly that are not deriving from `DiagnosticAnalyzer`, but may be accessed from diagnostic analyzers. These types might have helper methods that access types from Workspaces assemblies.

heejaechang commented on src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAPIUsageAnalyzer.cs:
```diff
@@ -0,0 +1,200 @@
+                            Debug.Assert(typeToProcess.ContainingAssembly == declaredType.ContainingAssembly);
+                            Debug.Assert(namedTypesToAccessedTypesMap.ContainsKey(typeToProcess));
+
+                            foreach (INamedTypeSymbol usedType in namedTypesToAccessedTypesMap[typeToProcess])
```
I am not sure relation between this and Getusedtype which set hasaccesstoworkspaceassembly stuff.

heejaechang replied:
outer loop will go through all analyzers anyway, so is this following types to improve diagnostic message?

mavasani replied:
Yes, I did not have this initially, but I found that the diagnostic message is then not very helpful. If analyzer type `A` accesses `B` which accesses something from Workspaces, say `C`, just stating that _change `A` to not access `C`_ is not very helpful. Now we state that change _`A` to not access `B` as it accesses `C`_

mavasani commented:
@AArnott 

> Use of extension methods defined in the workspaces assembly
Added test `IndirectlyAccessedType_ExtensionMethod_Diagnostic`

> Also how about type checks and casts--is that covered?
Added test `DirectlyAccessedType_CastAndTypeCheck_Diagnostic`

mavasani commented:
@sharwell Added unit test `IndirectlyAccessedType_StaticMember_Diagnostic` to cover your scenario.

mavasani commented:
Ping.. I believe the analyzer diagnostic is not tied to any specific template design and just points out a pure API usage violation. If in future we decide to split the template, we can change the diagnostic if required.

sharwell commented:
Approving as a safeguard against unintentional errors when packages are combined, though my strong and continued recommendation is for all analyzers to be spit with analyzers in one assembly and code fixes in another.

mavasani commented:
Thanks @sharwell

AArnott commented:
So where can we pick up analyzers with this change?

mavasani commented:
@AArnott Verified diagnostic in https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.Analyzers/2.6.1-beta1-62706-01

AArnott commented:
it works!
